File: backend/oasis_worker/src/oasis_worker/research_survey_contracts.py
"""Worker contracts for native single-context research surveys."""

from datetime import datetime
from typing import Annotated, Literal
from uuid import UUID

from pydantic import Field, StringConstraints

from oasis_worker.contracts import RequiredText, Sha256, StrictModel
from oasis_worker.semantic_contracts import SemanticPersona

RESEARCH_SURVEY_PROMPT_VERSION = "sandowl-research-survey/v1"
RESEARCH_SURVEY_RUNNER_VERSION = "1.0.0"
RESEARCH_SURVEY_TOOL_NAME = "submit_research_observation"


class ResearchSurveyRuntimeConfig(StrictModel):
    api_key: Annotated[str, StringConstraints(min_length=1, max_length=8192, strict=True)] = Field(
        repr=False
    )
    base_url: Annotated[str, StringConstraints(min_length=1, max_length=2000, strict=True)] = Field(
        repr=False
    )
    model_name: Annotated[RequiredText, Field(max_length=200)]
    config_sha256: Sha256
    prompt_schema_version: Literal["sandowl-research-survey/v1"]


class ResearchSurveyContext(StrictModel):
    id: UUID
    project_id: UUID
    run_id: UUID
    project_title: Annotated[RequiredText, Field(max_length=300)]
    research_question: Annotated[RequiredText, Field(max_length=2000)]
    simulation_requirement: Annotated[RequiredText, Field(max_length=4000)]
    initial_post: Annotated[RequiredText, Field(max_length=4000)]
    project_sha256: Sha256
    run_spec_sha256: Sha256
    cohort_id: UUID
    cohort_sha256: Sha256
    persona_count: Annotated[int, Field(ge=1, le=8)]
    model_name: Annotated[RequiredText, Field(max_length=200)]
    survey_config_sha256: Sha256
    survey_sha256: Sha256


class ClaimedResearchSurveyTrial(StrictModel):
    id: UUID
    created_at: datetime
    persona_position: Annotated[int, Field(ge=0, le=7)]
    persona_id: UUID
    persona_external_id: Annotated[
        str, StringConstraints(min_length=1, max_length=128, strict=True)
    ]
    persona_display_name: Annotated[RequiredText, Field(max_length=200)]
    persona_profile_sha256: Sha256
    trial_sha256: Sha256
    survey: ResearchSurveyContext
    persona: SemanticPersona


class ContextClarityAnswer(StrictModel):
    position: Literal[0]
    question_id: Literal["context_clarity"]
    type: Literal["likert"]
    value: Annotated[int, Field(strict=True, ge=1, le=5)]


class AttentionPriorityAnswer(StrictModel):
    position: Literal[1]
    question_id: Literal["attention_priority"]
    type: Literal["single_choice"]
    value: Literal["evidence", "process", "timing", "impact"]


class UnansweredQuestionAnswer(StrictModel):
    position: Literal[2]
    question_id: Literal["unanswered_question"]
    type: Literal["free_text"]
    value: Annotated[
        str, StringConstraints(min_length=1, max_length=2000, strip_whitespace=True, strict=True)
    ]


class ExtractedResearchSurveyResponse(StrictModel):
    answers: tuple[ContextClarityAnswer, AttentionPriorityAnswer, UnansweredQuestionAnswer]


class ResearchSurveySuccess(StrictModel):
    runner_version: Literal["1.0.0"]
    model_name: Annotated[RequiredText, Field(max_length=200)]
    survey_config_sha256: Sha256
    prompt_schema_version: Literal["sandowl-research-survey/v1"]
    answers: tuple[ContextClarityAnswer, AttentionPriorityAnswer, UnansweredQuestionAnswer]
    answers_sha256: Sha256

File: frontend/src/researchSurveyContracts.ts
import { z } from "zod";

import { getJson, postJson } from "./apiClient";
import { sha256DigestSchema } from "./mediaContracts";
import { parentProgressSchema, type ParentProgress } from "./parentProgress";

const endpoint = "/api/v2/research-surveys";
const id = z.string().uuid();
const text = z.string().trim().min(1);
const timestamp = z.string().datetime({ offset: true });
const status = z.enum(["queued", "running", "succeeded", "failed"]);
const projectRef = z.object({ id, title: text.max(300), research_question: text.max(2000), project_sha256: sha256DigestSchema }).strict();
const runRef = z.object({ id, simulation_requirement: text.max(4000), initial_post: text.max(4000), run_spec_sha256: sha256DigestSchema }).strict();
const cohortRef = z.object({ id, title: text.max(200), cohort_sha256: sha256DigestSchema, dataset_sha256: sha256DigestSchema, persona_count: z.number().int().min(1).max(8) }).strict();
const personaRef = z.object({ id, position: z.number().int().min(0).max(7), persona_id: text.max(128), display_name: text.max(200), profile_sha256: sha256DigestSchema }).strict();
const instrument = z.object({ schema_version: z.literal("single-context-observation/v1"), instrument_sha256: sha256DigestSchema, title: z.literal("Single-context observation"), description: text.max(4000) }).strict();
const answers = z.tuple([
  z.object({ position: z.literal(0), question_id: z.literal("context_clarity"), type: z.literal("likert"), value: z.number().int().min(1).max(5) }).strict(),
  z.object({ position: z.literal(1), question_id: z.literal("attention_priority"), type: z.literal("single_choice"), value: z.enum(["evidence", "process", "timing", "impact"]) }).strict(),
  z.object({ position: z.literal(2), question_id: z.literal("unanswered_question"), type: z.literal("free_text"), value: text.max(2000) }).strict(),
]);
const trial = z.object({ id, status, persona: personaRef, trial_sha256: sha256DigestSchema, created_at: timestamp, started_at: timestamp.nullable(), completed_at: timestamp.nullable(), result: z.object({ runner_version: z.literal("1.0.0"), model_name: text.max(200), survey_config_sha256: sha256DigestSchema, prompt_schema_version: z.literal("sandowl-research-survey/v1"), answers_sha256: sha256DigestSchema, answers }).strict().nullable(), error: z.object({ code: text.max(128), message: text.max(4000) }).strict().nullable() }).strict();
const summaryObject = z.object({ id, status, project: projectRef, run: runRef, cohort: cohortRef, trial_count: z.number().int().min(1).max(8), succeeded_trial_count: z.number().int().min(0).max(8), failed_trial_count: z.number().int().min(0).max(8), model_name: text.max(200), survey_config_sha256: sha256DigestSchema, prompt_schema_version: z.literal("sandowl-research-survey/v1"), instrument_schema_version: z.literal("single-context-observation/v1"), instrument_sha256: sha256DigestSchema, survey_sha256: sha256DigestSchema, created_at: timestamp }).strict();
export const researchSurveySummarySchema = summaryObject;
export const researchSurveyDetailSchema = summaryObject.extend({ instrument, trials: z.array(trial).min(1).max(8), aggregate: z.object({ succeeded_trial_count: z.number().int().min(0).max(8), failed_trial_count: z.number().int().min(0).max(8), context_clarity_mean: z.number().min(1).max(5).nullable(), attention_priority: z.object({ evidence: z.number().int().min(0).max(8), process: z.number().int().min(0).max(8), timing: z.number().int().min(0).max(8), impact: z.number().int().min(0).max(8) }).strict(), unanswered_questions: z.array(text.max(4000)).max(8), limitations: z.array(text.max(4000)).min(1) }).strict() }).strict();
export const researchSurveyReadinessSchema = z.object({ engine: z.literal("matraix-survey"), runner_version: z.literal("1.0.0"), survey_runtime_ready: z.boolean(), live_worker_count: z.number().int().nonnegative(), model_name: text.max(200).nullable(), survey_config_sha256: sha256DigestSchema.nullable(), prompt_schema_version: z.literal("sandowl-research-survey/v1").nullable(), instrument_schema_version: z.literal("single-context-observation/v1"), limitations: z.array(text.max(4000)).min(1) }).strict();
const response = z.object({ items: z.array(researchSurveySummarySchema), total: z.number().int().nonnegative() }).strict();
const createRequest = z.object({ research_project_id: id, research_simulation_run_id: id }).strict();

export type ResearchSurveySummary = z.infer<typeof researchSurveySummarySchema>;
export type ResearchSurveyDetail = z.infer<typeof researchSurveyDetailSchema>;
export type ResearchSurveyTrial = z.infer<typeof trial>;
export type ResearchSurveyReadiness = z.infer<typeof researchSurveyReadinessSchema>;
export const fetchResearchSurveys = (signal: AbortSignal): Promise<z.infer<typeof response>> => getJson(endpoint, response, signal);
export const fetchResearchSurvey = (surveyId: string, signal: AbortSignal): Promise<ResearchSurveyDetail> => getJson(`${endpoint}/${id.parse(surveyId)}`, researchSurveyDetailSchema, signal);
export const fetchResearchSurveyProgress = (surveyId: string, signal: AbortSignal): Promise<ParentProgress> => getJson(`${endpoint}/${id.parse(surveyId)}/progress`, parentProgressSchema, signal);
export const fetchResearchSurveyReadiness = (signal: AbortSignal): Promise<ResearchSurveyReadiness> => getJson(`${endpoint}/readiness`, researchSurveyReadinessSchema, signal);
export const createResearchSurvey = (projectId: string, runId: string, signal: AbortSignal): Promise<ResearchSurveyDetail> => postJson(endpoint, createRequest.parse({ research_project_id: projectId, research_simulation_run_id: runId }), researchSurveyDetailSchema, signal);
